cpp/src/arrow/util/logging.h

```diff
@@ -55,7 +55,12 @@ enum class ArrowLogLevel : int {
 };
 
 #define ARROW_LOG_INTERNAL(level) ::arrow::util::ArrowLog(__FILE__, __LINE__, level)
-#define ARROW_LOG(level) ARROW_LOG_INTERNAL(::arrow::util::ArrowLogLevel::ARROW_##level)
+// clang-format off
+#define ARROW_LOG(level) if (                                                       \
+  arrow::util::ArrowLog::IsLevelEnabled(arrow::util::ArrowLogLevel::ARROW_##level)) \
+  ARROW_LOG_INTERNAL(arrow::util::ArrowLogLevel::ARROW_##level)
+// clang-format on
```

Comment:
What if I write something like:
```cpp
if (x)
  ARROW_LOG(DEBUG) << "some message";
else
  // something else...;
```

Will the "something else" branch be executed if x == 0 ?

Reply:
@pitrou The problem is that we often use [`ARROW_LOG(WARNING) << "some message"`](https://github.com/apache/arrow/blob/master/cpp/src/plasma/plasma.cc#L45). If there is a branch or braces, compilation error will occur.

Reply:
That doesn't answer the question, though. We shouldn't risk losing correctness over performance. If logging becomes expensive, it's always possible to add the `IsLevelEnabled` check manually in critical places.

Reply:
For that question, if `x == 0`, the else branch will be executed and the if branch will not be executed.However, if `x != 0` and the log level is high, there is no output from `ARROW_LOG(DEBUG)` but the operator `<<` is executed.
Currently, there is not perf problem in using this log. I will revert the change in the macro definition but keep the function `IsLevelEnabled`.